### ISSUE TITLE
potential mis-using of messageValidationVistior of ServerFactoryContext

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#31073]
*Title*: *potential mis-using of messageValidationVistior of ServerFactoryContext*

*Description*:

`ServerFactoryContext`, `FactoryContext`, `ClusterFactoryContext` provided `messageValidationVistior()` to validate xDS configuration.

For the `ServerFactoryContext`, dynamic visitor will be provided if the server is initialized.
And for the `FactoryContext` and `ClusterFactoryContext`, dynamic visitor will be provided if the related listeners/clusters are added by the xDS API.

IMO, the `messageValidationVistior()` should be selected depends on the scope of xDS configuration. For example, embedded route config should use the visitor provided by the `FactoryContext` and dynamic route config should always used dynamic visitor.

But in the current code base, there are lots of extensions don't obey this rule. For example, the stateful session filter (my bad), the `createFilterFactoryFromProtoWithServerContext` API, etc.

Seems this also require a complete checking and cleanup.

**Comment by wbpcode** [comment]
cc @alyssawilk @mattklein123

**Comment by alyssawilk** [comment]
yeah last time I did factory validation clean up I was very confused about what validation code was supposed to be used where because it seemed pretty inconsistent.

**Comment by htuch** [comment]
Yep, I think this is a case where inconsistent use has spread via imitation in the code base. There's two things we could do:
1. Clean up this code base wide based on the rubric that @wbpcode is hinting at (not sure I 100% understand everything in the OP comment but I agree with "For example, embedded route config should use the visitor provided by the FactoryContext and dynamic route config should always used dynamic visitor.").
2. Ask if we want to go back and revisit the whole "different validation for static vs. dynamic" requirement. If we did, we could massively simplify.

As I remember it, the original concern was that folks who were hand crafting static configs wanted the ability to get early feedback on typo-ed or unsupported field names, which might be treated as unknown fields and otherwise ignored. But dynamic config from an xDS server might validly use unknown fields, since the server might be more recent than client.

Arguably anyone who wants strict validation of unknown fields in the bootstrap could just run Envoy in "validation" mode with a universal strict flag and then regular Envoy serving operates with permissive proto support. This might still not be particularly great from UX perspective, but just want to throw it out there as a way to reduce code complexity.

**Comment by wbpcode** [comment]
> Ask if we want to go back and revisit the whole "different validation for static vs. dynamic" requirement. If we did, we could massively simplify.

I like this. :) In most of cases, users get almost all config from from xDS. It's much simple to let users choose one mode for the configuration validation.

But not sure if we have a backward compatibility promise in the configuration validation. 🤔

**Comment by htuch** [comment]
How bad is the cleanup based on what you've seen so far? I think if it's a fairly mechanical set of fixes we could just bite the bullet. I don't actually think removing this support would take away any key capability, but folks would need to run boostraps via config validation mode when working with relaxed validation with dynamic config.

**Comment by adisuissa** [comment]
I don't have enough context on this, but would like to understand the motivation before suggesting a solution.

Assuming that the goal is to keep the static validation for hand-crafted-config, and dynamic validation for config-from-dynamic-sources, I wonder if the two can be unified into a single validator that decides whether to execute the static/dynamic flavor depending on what's calling it (i.e., which context it is executed with).

**Comment by htuch** [comment]
> I wonder if the two can be unified into a single validator that decides whether to execute the static/dynamic flavor depending on what's calling it (i.e., which context it is executed with).

Sure, but that's the underlying challenge, propagating the context.
